pex/cli/commands/lockfile/updater.py

```diff
@@ -0,0 +1,205 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import
+
+import logging
+import os
+from collections import OrderedDict
+from contextlib import contextmanager
+
+from pex.cli.commands.lockfile import create
+from pex.commands.command import Error, ResultError, try_
+from pex.pep_503 import ProjectName
+from pex.resolve.locked_resolve import LockConfiguration, LockedRequirement, LockedResolve, Version
+from pex.resolve.requirement_configuration import RequirementConfiguration
+from pex.resolve.resolver_configuration import PipConfiguration
+from pex.resolve.target_configuration import TargetConfiguration
+from pex.sorted_tuple import SortedTuple
+from pex.third_party.pkg_resources import Requirement
+from pex.typing import TYPE_CHECKING
+from pex.util import named_temporary_file
+
+if TYPE_CHECKING:
+    import attr  # vendor:skip
+    from typing import Dict, Iterable, Iterator, Mapping, Optional, Union
+else:
+    from pex.third_party import attr
+
+
+logger = logging.getLogger(__name__)
+
+
+@attr.s(frozen=True)
+class VersionUpdate(object):
+    original = attr.ib()  # type: Version
+    updated = attr.ib()  # type: Version
+
+
+@attr.s(frozen=True)
+class LockUpdate(object):
+    updated_resolve = attr.ib()  # type: LockedResolve
+    updates = attr.ib()  # type: Mapping[ProjectName, Optional[VersionUpdate]]
+
+
+@attr.s(frozen=True)
+class LockUpdater(object):
+    """Updates a lockfile in whole or in part.
+
+    A lock updater with no updates specified just updates the whole lock file with the latest
+    available distributions that satisfy the locked resolve criteria.
+
+    More interestingly, a lock updater with a set of updates will attempt to alter just the
+    distributions for those updates. The update can be just a plain project name in which case the
+    latest compatible distribution for that project will be used. The update can also be a
+    full-fledged constraint with a specifier, in which case a distribution matching the constraint
+    and still satisfying the rest of the locked resolve will be searched for. This facility allows
+    for targeted upgrades and downgrades of individual projects in a lock that maintain the
+    integrity of the locked resolve.
+
+    When a set of updates is specified, the updater ensures all other project distributions remain
+    unchanged.
+    """
+
+    @classmethod
+    def create(
+        cls,
+        requirements,  # type: Iterable[Requirement]
+        constraints,  # type: Iterable[Requirement]
+        updates,  # type: Iterable[Requirement]
```

Comment:
Is it assumed that `updates` are always a (fuzzy) subset of `requirements`? Or will this do the right thing if given new requirements that did not exist in the set before?

Reply:
I'm not sure what the right thing is. Right now, disjoint updates will just go unused. The update mechanism is just to re-run the resolve using the original requirements + a full set of constraints covering the previous lock pinned transitive closure with an override for each `-p`. So if a `-p` is not in the original transitive closure, it still get written into the constraints file; it will just never be used in the course of the resolve.

Reply:
That seems fine, although probably worth documenting in `-p` that values for requirements not in the lockfile will be ignored.

Reply:
Ok, I used the precedent set by cargo here and now fail fast.